# Preparation

Preparation biasa dilakukan untuk mempersiapkan data sebelum masuk dalam tahap pemodelan. <br>
Berikut adalah tahapan yang akan dilalui pada data `SC_HW1_bank_data.csv` sebelum tahap pemodelan :
1. Import Library
2. Input Dataset
3. Preprocessing
4. Train-Test Split

## Import Library

In [1]:
import pandas as pd
import numpy as np

## Input Dataset

In [2]:
#Membaca data dan memasukkannya ke dalam bentuk Pandas Dataframe
df = pd.read_csv('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv')
df.head()

RowNumber  CustomerId Geography  Gender  Age  Tenure    Balance  \
0          1    15634602    France  Female   42       2       0.00   
1          2    15647311     Spain  Female   41       1   83807.86   
2          3    15619304    France  Female   42       8  159660.80   
3          4    15701354    France  Female   39       1       0.00   
4          5    15737888     Spain  Female   43       2  125510.82   

   NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0              1          1               1        101348.88       1  
1              1          0               1        112542.58       0  
2              3          1               0        113931.57       1  
3              2          0               0         93826.63       0  
4              1          1               1         79084.10       0

In [3]:
#Jalankan code untuk mengecek nama kolom yang tersedia
df.columns

Index(['RowNumber', 'CustomerId', 'Geography', 'Gender', 'Age', 'Tenure',
       'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited'],
      dtype='object')

In [4]:
#Hilangkan kolom yang dirasa tidak relevan dengan model (contoh: ID). None dapat diisi dengan nama-nama kolom yang akan digunakan.
#Contoh df = df[['X1','X2', 'Y']].copy()

#START CODE
df = df[['CustomerId', 'Geography', 'Gender', 'Age', 'Tenure',
       'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited']].copy()

In [5]:
df

CustomerId Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0       15634602    France  Female   42       2       0.00              1   
1       15647311     Spain  Female   41       1   83807.86              1   
2       15619304    France  Female   42       8  159660.80              3   
3       15701354    France  Female   39       1       0.00              2   
4       15737888     Spain  Female   43       2  125510.82              1   
...          ...       ...     ...  ...     ...        ...            ...   
9995    15606229    France    Male   39       5       0.00              2   
9996    15569892    France    Male   35      10   57369.61              1   
9997    15584532    France  Female   36       7       0.00              1   
9998    15682355   Germany    Male   42       3   75075.31              2   
9999    15628319    France  Female   28       4  130142.79              1   

      HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0             1               1        101348.88       1  
1             0               1        112542.58       0  
2             1               0        113931.57       1  
3             0               0         93826.63       0  
4             1               1         79084.10       0  
...         ...             ...              ...     ...  
9995          1               0         96270.64       0  
9996          1               1        101699.77       0  
9997          0               1         42085.58       1  
9998          1               0         92888.52       1  
9999          1               0         38190.78       0  

[10000 rows x 11 columns]

## Preprocessing

In [6]:
#Lakukan One-Hot Encoder untuk data categorical, dengan fungsi pandas get_dummies

#START CODE
df = pd.get_dummies(df, columns=['Geography', 'Gender'])

In [7]:
df

CustomerId  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0       15634602   42       2       0.00              1          1   
1       15647311   41       1   83807.86              1          0   
2       15619304   42       8  159660.80              3          1   
3       15701354   39       1       0.00              2          0   
4       15737888   43       2  125510.82              1          1   
...          ...  ...     ...        ...            ...        ...   
9995    15606229   39       5       0.00              2          1   
9996    15569892   35      10   57369.61              1          1   
9997    15584532   36       7       0.00              1          0   
9998    15682355   42       3   75075.31              2          1   
9999    15628319   28       4  130142.79              1          1   

      IsActiveMember  EstimatedSalary  Exited  Geography_France  \
0                  1        101348.88       1                 1   
1                  1        112542.58       0                 0   
2                  0        113931.57       1                 1   
3                  0         93826.63       0                 1   
4                  1         79084.10       0                 0   
...              ...              ...     ...               ...   
9995               0         96270.64       0                 1   
9996               1        101699.77       0                 1   
9997               1         42085.58       1                 1   
9998               0         92888.52       1                 0   
9999               0         38190.78       0                 1   

      Geography_Germany  Geography_Spain  Gender_Female  Gender_Male  
0                     0                0              1            0  
1                     0                1              1            0  
2                     0                0              1            0  
3                     0                0              1            0  
4                     0                1              1            0  
...                 ...              ...            ...          ...  
9995                  0                0              0            1  
9996                  0                0              0            1  
9997                  0                0              1            0  
9998                  1                0              0            1  
9999                  0                0              1            0  

[10000 rows x 14 columns]

In [8]:
#Pisahkan mana X (feature) dengan Y,
#Y adalah kolom "Exited"

#START CODE
X = df.drop(['Exited'], axis=1)
y = df['Exited']

In [9]:
X

CustomerId  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0       15634602   42       2       0.00              1          1   
1       15647311   41       1   83807.86              1          0   
2       15619304   42       8  159660.80              3          1   
3       15701354   39       1       0.00              2          0   
4       15737888   43       2  125510.82              1          1   
...          ...  ...     ...        ...            ...        ...   
9995    15606229   39       5       0.00              2          1   
9996    15569892   35      10   57369.61              1          1   
9997    15584532   36       7       0.00              1          0   
9998    15682355   42       3   75075.31              2          1   
9999    15628319   28       4  130142.79              1          1   

      IsActiveMember  EstimatedSalary  Geography_France  Geography_Germany  \
0                  1        101348.88                 1                  0   
1                  1        112542.58                 0                  0   
2                  0        113931.57                 1                  0   
3                  0         93826.63                 1                  0   
4                  1         79084.10                 0                  0   
...              ...              ...               ...                ...   
9995               0         96270.64                 1                  0   
9996               1        101699.77                 1                  0   
9997               1         42085.58                 1                  0   
9998               0         92888.52                 0                  1   
9999               0         38190.78                 1                  0   

      Geography_Spain  Gender_Female  Gender_Male  
0                   0              1            0  
1                   1              1            0  
2                   0              1            0  
3                   0              1            0  
4                   1              1            0  
...               ...            ...          ...  
9995                0              0            1  
9996                0              0            1  
9997                0              1            0  
9998                0              0            1  
9999                0              1            0  

[10000 rows x 13 columns]

In [10]:
y

0       1
1       0
2       1
3       0
4       0
       ..
9995    0
9996    0
9997    1
9998    1
9999    0
Name: Exited, Length: 10000, dtype: int64

In [11]:
#Lakukan Scaler dan/atau Noermalisasi jika diperlukan
from sklearn.preprocessing import MinMaxScaler

#START CODE
scaler = MinMaxScaler()
X_transform = scaler.fit_transform(X)

In [12]:
print(X_transform[:5])

[[0.27561613 0.32432432 0.2        0.         0.         1.
  1.         0.50673489 1.         0.         0.         1.
  0.        ]
 [0.32645436 0.31081081 0.1        0.33403148 0.         0.
  1.         0.56270874 0.         0.         1.         1.
  0.        ]
 [0.21442143 0.32432432 0.8        0.63635718 0.66666667 1.
  0.         0.56965435 1.         0.         0.         1.
  0.        ]
 [0.54263588 0.28378378 0.1        0.         0.33333333 0.
  0.         0.46912005 1.         0.         0.         1.
  0.        ]
 [0.68877831 0.33783784 0.2        0.50024622 0.         1.
  1.         0.39540036 0.         0.         1.         1.
  0.        ]]


In [13]:
#Ini digunakan jika dilakukan scaler/Normalisas. Jika tidak, code ini bisa dilewat dan diganti dengan code yang ada di dalam komen
X_transform = pd.DataFrame(X_transform, columns = X.columns)
#X_transform = X.copy()

## Train-Test Split

In [14]:
#Split menjadi train dan test dengan test_size 25%
#Tidak perlu mengubah code ini

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_transform,y,test_size = 0.25,random_state = 123)

# Modeling

## Model1

### Soal :
Jelaskan secara Singkat Model pertama yang digunakan!

pada model 1 saya menggunakan metode logistic regression

In [15]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


#START CODE
model1 = LogisticRegression()
params = {
    "tol": [0.1, 0.01, 0.001],
    'C': [0.5, 1.0, 1.5, 2.0]
}
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model1,
             param_grid= params,
             scoring = 'accuracy',
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid.fit(X_train,y_train)
grid.best_params_

{'C': 0.5, 'tol': 0.1}

In [17]:
#lakukan evaluasi dengan beberapa maetric di bawah ini
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("Confusion Matrix:")
print(confusion_matrix(y_test,y_pred))
print("Accuracy Score:")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.82      0.97      0.89      1983
           1       0.65      0.20      0.30       517

    accuracy                           0.81      2500
   macro avg       0.74      0.59      0.60      2500
weighted avg       0.79      0.81      0.77      2500

Confusion Matrix:
[[1929   54]
 [ 415  102]]
Accuracy Score:
0.8124


## Model2
### Soal :
Jelaskan secara Singkat Model ke-2 yang digunakan!

Pada model 2 saya menggunakan metode decision tree

In [18]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

from sklearn.tree import DecisionTreeClassifier
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


#START CODE
model2 = DecisionTreeClassifier()
params = {
    "criterion": ["gini", "entropy"],
    "max_depth": [None, 10, 20, 30],
    "min_samples_split": [2, 5, 10],
    "min_samples_leaf": [1, 2, 4]
}
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model2,
             param_grid= params,
             scoring = 'accuracy',
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid.fit(X_train,y_train)
grid.best_params_

{'criterion': 'gini',
 'max_depth': 10,
 'min_samples_leaf': 2,
 'min_samples_split': 10}

In [20]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("Confusion Matrix:")
print(confusion_matrix(y_test,y_pred))
print("Accuracy Score:")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.87      0.95      0.91      1983
           1       0.70      0.48      0.57       517

    accuracy                           0.85      2500
   macro avg       0.79      0.71      0.74      2500
weighted avg       0.84      0.85      0.84      2500

Confusion Matrix:
[[1878  105]
 [ 271  246]]
Accuracy Score:
0.8496


## Model3
### Soal :
Jelaskan secara Singkat Model ke-3 yang digunakan!

Pada model 3 saya menggunakan metode SVM

In [21]:
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV

# Inisialisasi model SVM
model3 = SVC()
params = {
    "C": [0.1, 1, 10],  # Parameter biaya (regularisasi)
    "kernel": ["linear", "poly", "rbf", "sigmoid"],  # Jenis kernel
    "degree": [2, 3, 4],  # Degree (hanya berlaku untuk kernel "poly")
}

# Lakukan GridSearchCV untuk mencari hyperparameter terbaik
grid = GridSearchCV(
    estimator=model3,
    param_grid=params,
    scoring='accuracy',
    n_jobs=10,
    cv=10
)

grid.fit(X_train, y_train)
best_params = grid.best_params_

# Lakukan prediksi menggunakan model terbaik
y_pred = grid.predict(X_test)

# Evaluasi hasil
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix

print(classification_report(y_test, y_pred))
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("Accuracy Score:", accuracy_score(y_test, y_pred))


              precision    recall  f1-score   support

           0       0.86      0.98      0.92      1983
           1       0.84      0.39      0.53       517

    accuracy                           0.86      2500
   macro avg       0.85      0.69      0.73      2500
weighted avg       0.86      0.86      0.84      2500

Confusion Matrix:
[[1945   38]
 [ 315  202]]
Accuracy Score: 0.8588


## Tarik Kesimpulan Model Mana yang terbaik beserta alasannya

Berdasarkan hasil evaluasi yang telah dilakukan didapatkan Model 1 (Logistic Regression) dengan akurasi: 0.8124, Model 2 (Decision Tree) dengan akurasi: 0.8448, dan Model 3 (Support Vector Machine - SVM) dengan akurasi: 0.8588. Dapat ditarik kesimpulan bahwa Model SVM memiliki akurasi yang lebih tinggi daripada Model 1 (Logistic Regression) dan Model 2 (Decision Tree).